internal/job/shell/shell_test.go

```diff
@@ -361,7 +357,107 @@ func newShellForTest(t *testing.T) *shell.Shell {
 	return sh
 }
 
+func TestRunWithoutPromptWithOlfactor(t *testing.T) {
+	t.Parallel()
+
+	for _, test := range []struct {
+		name        string
+		smell       string
+		expectSmell bool
+		command     []string
+		output      string
+		exitCode    int
+	}{
+		{
+			name:        "smells_nothing_when_no_error",
+			smell:       "hi",
+			expectSmell: true,
+			command:     []string{"bash", "-ec", "echo hi"},
+			output:      "hi\n",
+			exitCode:    0,
+		},
+		{
+			name:        "smells_stdout",
+			smell:       "hi",
+			expectSmell: true,
+			command:     []string{"bash", "-ec", "echo hi; false"},
+			output:      "hi\n",
+			exitCode:    1,
+		},
+		{
+			name:        "smells_stderr",
+			smell:       "hi",
+			expectSmell: true,
+			command:     []string{"bash", "-ec", "echo hi >&2; false"},
+			output:      "hi\n",
+			exitCode:    1,
+		},
+		{
+			name:        "smells_infix",
+			smell:       "hi",
+			expectSmell: true,
+			command:     []string{"bash", "-ec", "echo lorem ipsum; echo hi; echo bye; false"},
+			output:      "lorem ipsum\nhi\nbye\n",
+			exitCode:    1,
+		},
+		{
+			name:        "smells_infix_inline",
+			smell:       "hi",
+			expectSmell: true,
+			command:     []string{"bash", "-ec", "echo lorem hipsum; echo bye; false"},
+			output:      "lorem hipsum\nbye\n",
+			exitCode:    1,
+		},
+		{
+			name:        "smells_partial",
+			smell:       "hi",
+			expectSmell: true,
+			command:     []string{"bash", "-ec", "echo hi, how are you; false"},
+			output:      "hi, how are you\n",
+			exitCode:    1,
+		},
+		{
+			name:        "no_smell",
+			smell:       "hi",
+			expectSmell: false,
+			command:     []string{"bash", "-ec", "echo bye, how were you; false"},
+			output:      "bye, how were you\n",
+			exitCode:    1,
+		},
+	} {
+		test := test
+		t.Run(test.name, func(t *testing.T) {
+			t.Parallel()
+
+			sh, err := shell.New()
+			assert.NilError(t, err)
+
+			out := &bytes.Buffer{}
+			sh.Writer = out
+			if err := sh.RunWithOlfactor(
+				context.Background(),
+				test.smell,
+				test.command[0],
+				test.command[1:]...,
+			); err == nil {
+				// if there is no error, we expect an 0 exit code
+				assert.Check(t, test.exitCode == 0)
+			} else if derr := new(shell.OlfactoryError); !errors.As(err, &derr) {
+				assert.Check(t, !test.expectSmell, "no smell detected, but %s was expected", test.smell)
+				assert.ErrorContains(t, err, "exit status 1")
+			} else {
+				// if there was a OlfactoryError, we expect the smell to match the test.smell
+				// and the exit code to match the test.exitCode
+				assert.ErrorContains(t, err, "error running command: exit status 1, detected: hi")
```

Comment:
I think the main source of complication here is the need to test the combination of success and error against expected success/failure and expected error. `assert` makes it shorter but it still needs three branches.

Would it simplify if the test cases were split into two or more tables?

Reply:
Yeah, I might do that in a follow up